lib/llvm-ast.js

```diff
@@ -0,0 +1,146 @@
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+export class LlvmAstParser {
+    constructor(compilerProps) {
+        this.maxAstLines = 500;
+        if (compilerProps) {
+            this.maxAstLines = compilerProps('maxLinesOfAst', this.maxAstLines);
+        }
+
+        this.spanTypes = {
+            NONE: 'none',
+            EMPTY: 'empty',
+            LEFT: 'left',
+            RIGHT: 'right',
+            FULL: 'full',
+        };
+    }
+
+    parseSpan(line) {
+        const spanRegex = /<((?:line|col)[\d ,:ceilno]+)>/;
+        const twoLineRegex = /line:(\d+)[,:].*line:(\d+)[:>]/;
+        const leftLineRegex = /^line:(\d+)/;
+        const rightLineRegex = / line:(\d+):/;
+
+        let m = line.match(spanRegex);
+        if (m) {
+            let span = m[1];
+            let ll = span.match(twoLineRegex);
+            if (ll) {
+                return { type : this.spanTypes.FULL, left : Number(ll[1]), right: Number(ll[2])};
+            }
+            let left = span.match(leftLineRegex);
+            if (left) {
+                return { type : this.spanTypes.LEFT, left : Number(left[1])};
+            }
+            let right = span.match(rightLineRegex);
+            if (right) {
+                return { type : this.spanTypes.RIGHT, right : Number(right[1])};
+            }
+            return { type : this.spanTypes.EMPTY };
+        }
+        return { type : this.spanTypes.NONE };
+    }
+
+    parseAndSetSourceLines(output) {
+        var lfrom, lto;
+        for (var line of output) {
+            let span = this.parseSpan(line.text);
+            switch(span.type) {
+                case this.spanTypes.NONE:
+                case this.spanTypes.EMPTY:
+                    break;
+                case this.spanTypes.LEFT:
+                    lfrom = span.left;
+                    lto = lfrom;
+                    break;
+                case this.spanTypes.RIGHT:
+                    lfrom = lto;
+                    lto = span.right;
+                    break;
+                case this.spanTypes.FULL:
+                    lfrom = span.left;
+                    lto = span.right;
+                    break;
+            }
+            if (span.type !== this.spanTypes.NONE) {
+                line.source = { from : lfrom, to : lto };
+            }
+        }
+    }
+
+    processAst(output) {
+        output = output.stdout;
+
+        // Top level decls start with |- or `-
+        const topLevelRegex = /^([`|])-/;
+
+        // Refers to the user's source file rather than a system header
```

Comment:
Thanks for these great comments btw. Makes it a lot easier on me!

Reply:
TBH, these comments were there before me, I just moved this function over to its own file. But I'll add a bunch of comments nevertheless.

Reply:
oh! Thanks - don't feel you have to :) Only if it genuinely clarifies, and doesn't seems like it';ll bit-rot.